x-pack/plugins/ml/common/types/common.ts

```diff
@@ -19,3 +19,15 @@ export function dictionaryToArray<TValue>(dict: Dictionary<TValue>): TValue[] {
 export type DeepPartial<T> = {
   [P in keyof T]?: DeepPartial<T[P]>;
 };
+
+export type DeepReadonly<T> = T extends Array<infer R>
+  ? ReadonlyArray<DeepReadonly<T>>
+  : T extends Function
+  ? T
+  : T extends object
+  ? DeepReadonlyObject<T>
+  : T;
```

Comment:
The styleguide speaks out against multiple ternaries here https://github.com/elastic/kibana/blob/master/STYLEGUIDE.md#only-use-ternary-operators-for-small-simple-code

Personally I think it's fine to leave it as is, but maybe you can add a comment that explains the nesting at what each level does?

Reply:
Yeah I'm aware of multiple ternaries in kibana, just used an example from https://www.typescriptlang.org/docs/handbook/advanced-types.html